staging/src/k8s.io/client-go/discovery/content_type.go

```diff
@@ -0,0 +1,124 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"mime"
+	"reflect"
+	"sort"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	AcceptV1            = runtime.ContentTypeJSON
+	AcceptV2Beta1Params = sortedContentTypeParams{
+		params: map[string]string{
+			"g":  "apidiscovery.k8s.io",
+			"v":  "v2beta1",
+			"as": "APIGroupDiscoveryList",
+		},
+	}
+	// Aggregated discovery content-type (v2beta1). NOTE: content-type parameters
+	// MUST be ordered (g, v, as) for server in "Accept" header (BUT we are resilient
+	// to ordering when comparing returned values in "Content-Type" header).
+	AcceptV2Beta1 = runtime.ContentTypeJSON + ";" + AcceptV2Beta1Params.String()
+	// Prioritize aggregated discovery by placing first in the order of discovery accept types.
+	acceptDiscoveryFormats = AcceptV2Beta1 + "," + AcceptV1
+)
+
+// sortedContentTypeParams encapsulated content type parameters, where
+// the discovery content-type parameters will be ordered when printed,
+// but where parameter ordering will not matter when comparing.
+type sortedContentTypeParams struct {
+	params map[string]string
+}
+
+// sortedKeys returns content-type parameters sorted so keys
+// "g" > "v" > "as". No deterministic ordering if keys are not
+// discovery content-type parameters {"g", "v", "as"}.
+func (p *sortedContentTypeParams) sortedKeys() []string {
+	keys := make([]string, len(p.params))
+	for key := range p.params {
+		keys = append(keys, key)
+	}
+	// Ordering: "g" > "v" > "as"
+	sort.Slice(keys, func(i, j int) bool {
+		if strings.EqualFold(keys[i], "g") {
+			return true
+		} else if strings.EqualFold(keys[i],"as") {
+			return false
+		} else if strings.EqualFold(keys[i], "v") && strings.EqualFold(keys[j], "as") {
+			return true
+		} else {
+			return false
+		}
+	})
+	return keys
+}
+
+// String prints the content-type params using the sortedKeys array to
+// print the params in the order of the "sortedKeys" field.
+func (p *sortedContentTypeParams) String() string {
+	keyValues := []string{}
+	for _, key := range p.sortedKeys() {
+		if value, ok := p.params[key]; ok {
+			keyValues = append(keyValues, key+"="+value)
+		}
+	}
+	return strings.Join(keyValues, ";")
+}
+
+// Equals returns true if the passed parameters map is the same as the
+// receivers parameters. NOTE: the ordering of the parametes does NOT matter.
+func (p *sortedContentTypeParams) Equals(params map[string]string) bool {
+	return reflect.DeepEqual(p.params, params)
```

Comment:
doesn't this still break if a charset param is appended, as in the original report?